### PR TITLE
refactor: removed uncessary panics in zetaclientd process

### PR DESCRIPTION
**Pull request by ws4charlie**
# Description

1. Removed all panics in `zetaclientd` process, leaving the only one in `start.go` when `zetaclientd` fails to start telemetry server. 
2. Removed duplicate code in the `ObserveInTx` function of bitcoin.

Closes: [2024](https://github.com/zeta-chain/node/issues/2024)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [x] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: zetaclient/chains/bitcoin/observer/observer.go [hunk 1]
```diff
@@ -344,34 +344,24 @@ func (ob *Observer) Stop() {
...
 	atomic.StoreInt64(&ob.lastBlock, height)
 }
 
 func (ob *Observer) GetLastBlockHeight() int64 {
-	height := atomic.LoadInt64(&ob.lastBlock)
-	if height < 0 {
-		panic("lastBlock is negative")
-	}
-	return height
+	return atomic.LoadInt64(&ob.lastBlock)
```
Comment (ws4charlie) [inline zetaclient/chains/bitcoin/observer/observer.go:351]: height already checked before setting

**Review by lumtis — COMMENTED** [review]
We have this test in zetacore_context:
```
		assertPanic(t, func() {
			zetaContext.GetBTCChainParams()
		}, "BTCChain is missing for chainID 0")
```
Can't this be removed now? Do we still get panic on wrong Bitcoin params ?

File: cmd/zetacore_utils/main.go [hunk 2]
```diff
@@ -104,13 +107,14 @@ func main() {
 		output, err := cmd.CombinedOutput()
 		if err != nil {
 			fmt.Println(cmd.String())
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply cmd/zetacore_utils/main.go:109]: Not sure who will be the users of this CLI. Looks like the validator could use this for token distribution. I keep this log print untouched so it shows the exact (assembled) command that fails. The users in this case get more detailed info for their own troubleshooting.

File: zetaclient/chains/bitcoin/observer/inbound.go [hunk 3]
```diff
@@ -112,46 +112,6 @@ func (ob *Observer) ObserveInTx() error {
 		// filter incoming txs to TSS address
 		tssAddress := ob.Tss.BTCAddress()
 
-		if len(res.Block.Tx) > 1 {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/bitcoin/observer/inbound.go:115]: It was on [this line](https://github.com/zeta-chain/node/blob/remove-zetaclient-panics/zetaclient/chains/bitcoin/observer/inbound.go#L108)

File: zetaclient/chains/bitcoin/observer/live_test.go [hunk 4]
```diff
@@ -416,7 +416,7 @@ BLOCKLOOP:
 					continue BLOCKLOOP // retry the block
 				}
 				if senderAddr != mpvin.Prevout.ScriptpubkeyAddress {
-					panic(fmt.Sprintf("block %d, tx %s, vout %d: want %s, got %s\n", bn, vin.Txid, vin.Vout, mpvin.Prevout.ScriptpubkeyAddress, senderAddr))
+					t.Errorf("block %d, tx %s, vout %d: want %s, got %s\n", bn, vin.Txid, vin.Vout, mpvin.Prevout.ScriptpubkeyAddress, senderAddr)
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/bitcoin/observer/live_test.go:419]: The `t.Errorf` will fail and terminate the test. It is equal to a return.

File: zetaclient/keys/keys.go [hunk 5]
```diff
@@ -95,7 +98,7 @@ func (k *Keys) GetSignerInfo() *ckeys.Record {
 	signer := GetGranteeKeyName(k.signerName)
 	info, err := k.kb.Key(signer)
 	if err != nil {
-		panic(err)
+		return nil
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/keys/keys.go:101]: The method `GetSignerInfo` returns a pointer of `*ckeys.Record`. This line returns a `nil` to indicate error and it's tested in `TestGetSignerInfo`.

File: zetaclient/zetacore/client.go [hunk 6]
```diff
@@ -153,7 +156,11 @@ func (c *Client) SetAccountNumber(keyType authz.KeyType) {
 		c.logger.Error().Err(err).Msg("fail to get context")
 		return
 	}
-	address := c.keys.GetAddress()
+	address, err := c.keys.GetAddress()
+	if err != nil {
+		c.logger.Error().Err(err).Msg("fail to get address")
+		return
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/zetacore/client.go:162]: Yes indeed. It's better to change the method to return error and let `zetaclientd` to fail in `start.go`

File: zetaclient/chains/bitcoin/observer/observer.go [hunk 7]
```diff
@@ -344,34 +344,24 @@ func (ob *Observer) Stop() {
 }
 
 func (ob *Observer) SetLastBlockHeight(height int64) {
-	if height < 0 {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/bitcoin/observer/observer.go:347]: Now I removed the `if height < 0` in the method `SetLastBlockHeightScanned` as well. The `height < 0` was already checked before setting. The getter and setter look straightforward now. See commit https://github.com/zeta-chain/node/pull/2210/commits/4286a09db49991d18ec84be9b0a5fb73d1bdafac

File: zetaclient/zetacore/client.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/zetacore/client.go]: https://github.com/zeta-chain/node/pull/2210/commits/4286a09db49991d18ec84be9b0a5fb73d1bdafac

File: cmd/zetaclientd/p2p_diagnostics.go [hunk 8]
```diff
@@ -151,68 +151,67 @@ func RunDiagnostics(startLogger zerolog.Logger, peers p2p.AddrList, hotkeyPk cry
...
+			nr, err := stream.Read(buf)
+			if err != nil {
+				startLogger.Error().Err(err).Msgf("fail to read from stream to peer %s", peer)
 				err = stream.Close()
 				if err != nil {
 					startLogger.Warn().Err(err).Msgf("fail to close stream to peer %s", peer)
 				}
+				continue
+			}
+			//startLogger.Debug().Msgf("read %d bytes", nr)
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply cmd/zetaclientd/p2p_diagnostics.go]: https://github.com/zeta-chain/node/pull/2210/commits/4286a09db49991d18ec84be9b0a5fb73d1bdafac

File: zetaclient/zetacore/client.go [hunk 9]
```diff
@@ -174,10 +182,12 @@ func (c *Client) WaitForCoreToCreateBlocks() {
 		retryCount++
 		c.logger.Debug().Msgf("Failed to get latest Block , Retry : %d/%d", retryCount, DefaultRetryCount)
 		if retryCount > ExtendedRetryCount {
-			panic(fmt.Sprintf("Zetacore is not ready, waited for %d seconds", DefaultRetryCount*DefaultRetryInterval))
+			c.logger.Error().Msgf("Zetacore is not ready, waited for %d seconds", DefaultRetryCount*DefaultRetryInterval)
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/zetacore/client.go]: https://github.com/zeta-chain/node/pull/2210/commits/4286a09db49991d18ec84be9b0a5fb73d1bdafac

File: zetaclient/zetacore/client.go [hunk 10]
```diff
@@ -163,7 +170,8 @@ func (c *Client) SetAccountNumber(keyType authz.KeyType) {
 	c.seqNumber[keyType] = seq
 }
 
-func (c *Client) WaitForCoreToCreateBlocks() {
+// WaitForZetacoreToCreateBlocks returns true if zetacore is ready to create blocks
+func (c *Client) WaitForZetacoreToCreateBlocks() bool {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/zetacore/client.go]: https://github.com/zeta-chain/node/pull/2210/commits/4286a09db49991d18ec84be9b0a5fb73d1bdafac

**Comment by ws4charlie** [comment]
> We have this test in zetacore_context:
> 
> ```
> 		assertPanic(t, func() {
> 			zetaContext.GetBTCChainParams()
> 		}, "BTCChain is missing for chainID 0")
> ```
> 
> Can't this be removed now? Do we still get panic on wrong Bitcoin params ?

removed in https://github.com/zeta-chain/node/pull/2210/commits/4286a09db49991d18ec84be9b0a5fb73d1bdafac

**Review by lumtis — APPROVED** [review]

**Pull request merged**